neps/nep-0366.md

```diff
@@ -0,0 +1,148 @@
+ - This transaction is processed normally, by creating a receipt with copy of the all the actions into the `Receipt`.
+ - When processing `DelegateAction` a number of checks are done (see below), mainly `signature` matching user account's key.
+ - When such `Receipt` with valid `DelegateAction` in actions arrives to the user's account it gets executed. The executed means creation of a new Receipt with `receiver_id: AccountId`, `actions: Action` matching `receiver_id` and `actions` inside `DelegateAction`. 
+ - The new `Receipt` looks like normal receipt that would have been originating from user's account, with `predeccessor_id` equal to user's account.
+
+### DelegateAction
+
+Delegate action allows for an account to initiate a batch of actions on behalf of the receiving account, allowing to proxy actions. This is used in implementation of meta transactions.
+
+```rust
+pub struct DelegateAction {
+    /// Receiver of the delegated actions.
+    receiver_id: AccountId,
+    /// List of actions to be executed.
+    actions: Vec<Action>,
+ 
+    /// Public key that is used to sign this delegated action.
+    signer_pk: PublicKey,
+    /// Nonce is used to determine that the same delegate action is not sent twice by relayers and should match for given account's `signer_pk`. 
+    /// After this action is processed it will increment.
+    nonce: Nonce,
+    /// Signature of the originating user signing `DelegateActionMessage` formed out of the data in the `DelegateAction`.
+    signature: Signature,
+}
+```
+
+ Supporting a batch of `actions` means `DelegateAction` can initiate a complex steps like creating a new account and transferring funds, deploying a contract and executing an initialization function.
+
+***Validation***:
+To ensure that `DelegateAction` is correct, on receival the verification of signature is done: `verify_signature(hash(message), signer_pk, signature)`.
+
+The `message` is formed in the next format and must be signed by the user:
+```rust
+struct DelegateActionMessage {
+    /// If not None, should match the `predecessor_id` that have created `DelegateAction`.
+    sender_id: Option<AccountId>,
+    /// Matching the given account_id.
+    signer_id: AccountId,
+    /// Matching the `signer_pk` from `DelegateAction`.
+    public_key: PublicKey,
+    /// Nonce for the given `public_key` from `DelegateAction`.
+    nonce: Nonce,
+    /// Matching the `receiver_id` from `DelegateAction`.
+    receiver_id: AccountId,
+    /// Block hash to ensure validity.
+    /// Actions matching `actions` from `DelegateAction`.
+    actions: Vec<Action>,
+}
+```
+
+The next set of security concerns are addressed by this format:
+ - If format matches `Transaction`, the relayer can just send it directly, not receiving payment.
+ - Because set of actions can include pay back to the relayer (for example by paying in FT), the `sender_id` is added directly into the message to ensure that nobody else can send this message.
```

Comment:
```suggestion
 - Because the set of actions can include payment for the relayer (for example in FT), the `sender_id` is added directly into the message to ensure that nobody else can send this message.
```